File: test/fixtures/edge-cases/special-patterns.js
// Edge cases for JavaScript pattern matching

// Standard patterns
const standard = process.env.STANDARD_VAR;
const bracket = process.env['BRACKET_VAR'];

// Nested access
const nested = process.env.NESTED?.VAR;
const optional = process.env?.OPTIONAL_VAR;

// Destructuring
const { DATABASE_URL, API_KEY } = process.env;
const { PORT = 3000 } = process.env;

// Template literals
const url = `${process.env.BASE_URL}/api`;
const connection = `postgres://${process.env.DB_USER}:${process.env.DB_PASS}@${process.env.DB_HOST}`;

// Conditional access
const config = process.env.NODE_ENV === 'production' ? process.env.PROD_KEY : process.env.DEV_KEY;

// Function calls
const port = parseInt(process.env.PORT);
const enabled = Boolean(process.env.FEATURE_ENABLED);

// Object spread
const envConfig = { ...process.env };

// Array access (should not match)
const arr = ['test'];
const notEnv = arr.env.SHOULD_NOT_MATCH;

// Comments containing env vars (should not match)
// const commented = process.env.COMMENTED_VAR;
/* const multiline = process.env.MULTILINE_VAR; */

// String literals (should not match)
const str = "process.env.STRING_LITERAL";
const template = `process.env.TEMPLATE_LITERAL`;

// Dynamic access (should not match specific var)
const dynamic = process.env[varName];
const computed = process.env[`${prefix}_VAR`];

// Vite/import.meta.env patterns
const viteVar = import.meta.env.VITE_API_URL;
const viteBracket = import.meta.env['VITE_PUBLIC_KEY'];
const viteTemplate = `${import.meta.env.VITE_BASE_URL}/api`;

// Edge case: multiple on same line
const multi = process.env.VAR1 || process.env.VAR2 || process.env.VAR3;

// Edge case: chained access
const chained = process.env.CHAINED_VAR?.toString().toLowerCase();

// Edge case: in function parameters
function getConfig(url = process.env.DEFAULT_URL) {
  return url;
}

// Edge case: in class properties
class Config {
  apiKey = process.env.CLASS_API_KEY;
  static baseUrl = process.env.STATIC_BASE_URL;
}

// Edge case: in arrow functions
const getPort = () => process.env.ARROW_PORT;
const getHost = () => { return process.env.ARROW_HOST; };

// Edge case: in ternary
const value = condition ? process.env.TRUE_VAR : process.env.FALSE_VAR;

// Edge case: in logical operators
const fallback = process.env.PRIMARY || process.env.SECONDARY || 'default';
const check = process.env.CHECK && process.env.VERIFY;

// Edge case: in array/object literals
const config = {
  db: process.env.DB_URL,
  cache: process.env.CACHE_URL,
  ports: [process.env.HTTP_PORT, process.env.HTTPS_PORT]
};

File: test/fixtures/edge-cases/special-patterns.py
#!/usr/bin/env python3
"""Edge cases for Python pattern matching"""

import os
from os import environ, getenv

# Standard patterns
standard = os.environ['STANDARD_VAR']
get_method = os.getenv('GET_VAR')
environ_direct = environ['ENVIRON_VAR']
getenv_direct = getenv('GETENV_VAR')

# With defaults
with_default = os.getenv('WITH_DEFAULT', 'default_value')
with_none = os.getenv('WITH_NONE', None)
with_int = int(os.getenv('WITH_INT', '0'))

# Dictionary access
dict_access = os.environ.get('DICT_ACCESS')
dict_bracket = os.environ['DICT_BRACKET']

# In f-strings
url = f"{os.environ['BASE_URL']}/api"
connection = f"postgres://{os.getenv('DB_USER')}:{os.getenv('DB_PASS')}@{os.getenv('DB_HOST')}"

# In conditionals
if os.environ.get('FEATURE_ENABLED'):
    config = os.environ['PROD_CONFIG']
else:
    config = os.getenv('DEV_CONFIG', 'default')

# In function calls
port = int(os.getenv('PORT', '3000'))
enabled = bool(os.environ.get('ENABLED', False))

# In list/dict comprehensions
env_vars = {k: v for k, v in os.environ.items() if k.startswith('APP_')}

# Comments (should not match)
# standard_commented = os.environ['COMMENTED_VAR']
"""
multiline_commented = os.getenv('MULTILINE_VAR')
"""

# String literals (should not match)
string_literal = "os.environ['STRING_LITERAL']"
template = f"os.getenv('TEMPLATE_LITERAL')"

# Dynamic access (should not match specific var)
dynamic = os.environ[var_name]
computed = os.getenv(f"{prefix}_VAR")

# Multiple on same line
multi = os.getenv('VAR1') or os.getenv('VAR2') or os.getenv('VAR3')

# In function parameters
def get_config(url=os.getenv('DEFAULT_URL')):
    return url

# In class attributes
class Config:
    api_key = os.environ.get('CLASS_API_KEY')
    base_url = os.getenv('CLASS_BASE_URL')
    
    def __init__(self):
        self.db_url = os.environ['INIT_DB_URL']
        self.cache_url = os.getenv('INIT_CACHE_URL')

# In lambda functions
get_port = lambda: os.getenv('LAMBDA_PORT')
get_host = lambda: os.environ.get('LAMBDA_HOST', 'localhost')

# In ternary expressions
value = os.getenv('TRUE_VAR') if condition else os.getenv('FALSE_VAR')

# In logical operators
fallback = os.getenv('PRIMARY') or os.getenv('SECONDARY') or 'default'
check = os.getenv('CHECK') and os.getenv('VERIFY')

# In data structures
config = {
    'db': os.environ['DB_URL'],
    'cache': os.getenv('CACHE_URL'),
    'ports': [os.getenv('HTTP_PORT'), os.getenv('HTTPS_PORT')]
}

# Multiline
multiline_value = os.environ.get(
    'MULTILINE_VAR',
    'default_value'
)

# Method chaining
upper_value = os.getenv('UPPER_VAR', '').upper()
split_value = os.environ.get('SPLIT_VAR', '').split(',')
